services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloadResumeContext.java

```diff
@@ -0,0 +1,140 @@
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.multipart;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.utils.ToString;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * This class keep tracks of the state of a multipart download across multipart GET requests.
+ */
+@SdkInternalApi
+public class MultipartDownloadResumeContext {
+
+    /**
+     * Keeps track of complete parts in a list sorted in ascending order
+     */
+    private final SortedSet<Integer> completedParts;
+
+    /**
+     * Keep track of the byte index to the last byte of the last completed part
+     */
+    private Long bytesToLastCompletedParts;
+
+    /**
+     * The total number of parts of the multipart download.
+     */
+    private Integer totalParts;
+
+    /**
+     * The GetObjectResponse to return to the user.
+     */
+    private GetObjectResponse response;
+
+    public MultipartDownloadResumeContext() {
+        this(new TreeSet<>(), 0L);
+    }
+
+    public MultipartDownloadResumeContext(Collection<Integer> completedParts, Long bytesToLastCompletedParts) {
+        this.completedParts = new TreeSet<>(Validate.notNull(
+            completedParts, "completedParts must not be null"));
+        this.bytesToLastCompletedParts = Validate.notNull(
+            bytesToLastCompletedParts, "bytesToLastCompletedParts must not be null");
+    }
+
+    public List<Integer> completedParts() {
+        return Arrays.asList(completedParts.toArray(new Integer[0]));
+    }
+
+    public Long bytesToLastCompletedParts() {
+        return bytesToLastCompletedParts;
+    }
+
+    public void addCompletedPart(int partNumber) {
+        completedParts.add(partNumber);
+    }
+
+    public void addToBytesToLastCompletedParts(long bytes) {
+        bytesToLastCompletedParts += bytes;
+    }
+
+    public void totalParts(int totalParts) {
+        this.totalParts = totalParts;
+    }
+
+    public Integer totalParts() {
+        return totalParts;
+    }
+
+    public GetObjectResponse response() {
+        return this.response;
+    }
+
+    public void response(GetObjectResponse response) {
+        this.response = response;
+    }
+
+    /**
+     * Return the highest sequentially completed part, 0 means no parts completed
+     *
+     * @return
+     */
+    public int highestSequentialCompletedPart() {
+        if (completedParts.isEmpty() || completedParts.first() != 1) {
+            return 0;
+        }
+        if (completedParts.size() == 1) {
+            return 1;
+        }
+
+        int previous = completedParts.first();
```

Comment:
Can we add a comment here to mention this is to handle the case where parts are not downloaded in sequential order, took me a few minutes to figure out the intention.

This seems to be specific for sequential consumer. I wonder if we should move the logic to the consumer itself. For file based transfers, I'd assume we'll just pass the set to the transformers in the future.

Reply:
I see, I will update javadoc for the method to make to make it more clear. This logic is very specific to the set of completed parts, which is why I would keep it here. But I don't have a strong feeling about it if you really prefer to move it somewhere else.

Reply:
Let's just add a comment here.